src/raftstore/coprocessor/mod.rs

```diff
@@ -72,19 +71,15 @@ pub trait QueryObserver: Coprocessor {
     /// Hook to call before proposing write request.
     ///
     /// We don't propose read request, hence there is no hook for it yet.
-    fn pre_propose_query(
-        &self,
-        _: &mut ObserverContext<'_>,
-        _: &mut RepeatedField<Request>,
-    ) -> Result<()> {
+    fn pre_propose_query(&self, _: &mut ObserverContext<'_>, _: &mut [Request]) -> Result<()> {
```

Comment:
According to the scenario suggested in https://github.com/tikv/tikv/pull/5112/files#r305574153 , looks like `&mut Vec<Request>` would a better idea since it allows adding or removing requests in the array.

Reply:
The problem with doing that is that when we call these methods we pass an `&mut` reference to the either the `RepeatedField` or `Vec` inside the decoded message, so we can't just pass an `&mut Vec` here, because we can't get a view of a `RepeatedField` as a `Vec`. I don't think there is a straightforward way to solve this. Either we pass `&mut [...]` or we make major changes to the API. (In any case, it seems very bad that we are leaking an implementation detail like which protobuf codec we are using into a public API).

Reply:
Yeah. From my own perspective of view, according to the purpose, which is, minimizing future efforts when cherry picking PRs (which requires our code to support both Prost and rust-protobuf), leaving a Vec here looks to be fine because it is our best effort without losing flexibility. @BusyJay PTAL

Reply:
> leaving a Vec here looks to be fine

Do you mean `&mut [...]`?

Reply:
@breeswish I agree.

> Do you mean &mut [...]?

@nrc he means `&mut Vec<Request>`. You can get a view of a `RepeatedField` as a `Vec` by Into trait and From trait.

Reply:
As far as I can see, that doesn't work because the into/from methods must take ownership of the input and we only have access to a reference. Although a Vec can be converted to and from a RepeatedField, you can't convert a reference without cloning.

Reply:
You can do it by `take`.

Reply:
Oh nice, that is new; thanks for the idea! It works really well here.

PR incorporates this change now